### ISSUE TITLE
use luaffifb to make it work in places other than luajit

### ISSUE THREAD

**Issue by jcupitt** [libvips/lua-vips#9]
See https://github.com/facebook/luaffifb

**Comment by rolandlo** [comment]
So should `lua-vips` work with Lua 5.3 when `luaffifb` is installed? 
I did so and added a replacement for `bit.band` (bitwise and), which looks like the only thing required from other libraries.
However I get a segmentation fault after a short while:
```lua
Lua 5.3.6  Copyright (C) 1994-2020 Lua.org, PUC-Rio
> vips = require "vips"
> vips.Image
table: 0x558ea766a7a0
> im = vips.Image.new_from_file("dog.png")
Speicherzugriffsfehler (Speicherabzug geschrieben)
```

**Comment by jcupitt** [comment]
Hi @rolandlo, I think it would need some dev work to make this happen. I've not had time, but if you do, a PR would be great.

**Comment by rolandlo** [comment]
Seems I managed to get it working on Lua 5.3.6. At least the [example from the Readme](https://github.com/libvips/lua-vips#example) works when I rebase [my patches](https://github.com/libvips/lua-vips/compare/master...rolandlo:lua-vips:fix-lua-5.3) onto lua-vips version 1.1.9 (commit 4115e8eeec02077919cd883f30cbc59026515ee4). 

![grafik](https://github.com/libvips/lua-vips/assets/40485433/43ac8478-7177-46cb-ba7f-59ea6ab2028a)

The images created (`tiny.jpg` and `x.png`) look fine. For this to work I also had to add some [changes](https://github.com/Tekenlight/luaffifb_fork/compare/master...rolandlo:luaffifb:master) to [Tekenlight's fork of luaffifb](https://github.com/Tekenlight/luaffifb_fork).

So far I haven't had time to test further. Without rebasing on version 1.1.9 I get a segfault when executing 
```lua
image = image = image + {1,2,3}
``` 
from the example.

Let me point out that I don't have much experience with Lua and the Lua C API and have not worked with `lua-vips` before, so most likely my solution can and should be improved a lot.

The code changes are motivated by the following observations:

- Currently (on master) `lua-vips` uses `unpack` which should be `table.unpack` in Lua 5.3. So I replaced it by `unpack or table.unpack`
- Currently `lua-vips` compares `gv.gtype` (which is of type `cdata`) with Lua numbers like `gvalue.gbool_type`, so I used `tonumber` to convert `gv.gtype` into a Lua number
- Currently `luaffifb` (both on the Facebook repo and on Tekenlight's fork) doesn't support things like `voperation.call` (or more generally `<ffi-metatype>.<function name>`) since the `ctype_mt` doesn't support the `__index` metamethod. So I simply added `ctype_index` by copying the relevant part from `cdata_index`. That looks a bit hacky, but it works.

**Comment by jcupitt** [comment]
That's very cool!

Sorry, I have no free time right now :( After July I should have some time to look at this again.

**Comment by rolandlo** [comment]
A little update: I have used my patched versions of `lua-vips` and `luaffifb` to write a Xournal++ plugin, where images in a Xournal++ document can be manipulated in various ways, see https://github.com/xournalpp/xournalpp/pull/4884. It seems to work fine (using Lua 5.3.6 on Ubuntu 23.04).

**Comment by rolandlo** [comment]
One more update: 
- Rebased on master
- Fixed some segfaults related to comparing with `nil` instead of `ffi.NULL`
- All examples in the `example` directory except of `array.lua` work now. In `array.lua`, 
```lua
local ptype = ffi.typeof("typedef unsigned short int[$][?]", im:bands())
```
results in the error message:
```term
lua: array.lua:14: unknown type typedef on line 1
stack traceback:
	[C]: in function 'ffi.typeof'
	array.lua:14: in main chunk
	[C]: in ?
```
In `watermark.lua` I had to replace division `/` by integer division `//` in `text:replicate` (which requires integer arguments)

**Comment by rolandlo** [comment]
Fixed by #60.
